src/openxr/extensions/xr_fb_passthrough_extension_wrapper.cpp

```diff
@@ -0,0 +1,216 @@
+#include "xr_fb_passthrough_extension_wrapper.h"
+
+#include <gen/Viewport.hpp>
+
+using namespace godot;
+
+XRFbPassthroughExtensionWrapper *XRFbPassthroughExtensionWrapper::singleton = nullptr;
+
+XRFbPassthroughExtensionWrapper *XRFbPassthroughExtensionWrapper::get_singleton() {
+	if (!singleton) {
+		singleton = new XRFbPassthroughExtensionWrapper();
+	}
+
+	return singleton;
+}
+
+XRFbPassthroughExtensionWrapper::XRFbPassthroughExtensionWrapper() {
+	openxr_api = OpenXRApi::openxr_get_api();
+	request_extensions[XR_FB_PASSTHROUGH_EXTENSION_NAME] = &fb_passthrough_ext;
+	request_extensions[XR_FB_TRIANGLE_MESH_EXTENSION_NAME] = &fb_triangle_mesh_ext;
+}
+
+XRFbPassthroughExtensionWrapper::~XRFbPassthroughExtensionWrapper() {
+	cleanup();
+	OpenXRApi::openxr_release_api();
+}
+
+void XRFbPassthroughExtensionWrapper::cleanup() {
+	fb_passthrough_ext = false;
+	fb_triangle_mesh_ext = false;
+}
+
+void XRFbPassthroughExtensionWrapper::on_instance_initialized(const XrInstance instance) {
+	if (fb_passthrough_ext) {
+		Godot::print("Initializing passthrough extension...");
+		XrResult result = initialize_fb_passthrough_extension(instance);
+		if (!openxr_api->xr_result(result, "Failed to initialize fb_passthrough extension")) {
+			fb_passthrough_ext = false;
+		}
+	}
+
+	if (fb_triangle_mesh_ext) {
+		XrResult result = initialize_fb_triangle_mesh_extension(instance);
+		if (!openxr_api->xr_result(result, "Failed to initialize fb_triangle_mesh extension")) {
+			fb_triangle_mesh_ext = false;
+		}
+	}
+
+	if (fb_passthrough_ext) {
+		Godot::print("Registering as composition layer provider...");
+		openxr_api->register_composition_layer_provider(this);
+	}
+}
+
+bool XRFbPassthroughExtensionWrapper::is_passthrough_valid() {
+	return fb_passthrough_ext && passthrough_handle != XR_NULL_HANDLE && passthrough_layer != XR_NULL_HANDLE;
+}
+
+bool XRFbPassthroughExtensionWrapper::is_composition_passthrough_layer_ready() {
+	return fb_passthrough_ext && passthrough_handle != XR_NULL_HANDLE && composition_passthrough_layer.layerHandle != XR_NULL_HANDLE;
+}
+
+bool XRFbPassthroughExtensionWrapper::start_passthrough() {
+	if (!is_passthrough_valid()) {
+		return false;
+	}
+
+	// Start the passthrough feature
+	Godot::print("Starting passthrough feature...");
+	XrResult result = xrPassthroughStartFB(passthrough_handle);
+	if (!openxr_api->xr_result(result, "Failed to start passthrough")) {
+		stop_passthrough();
+		return false;
+	}
+
+	// Resume the passthrough layer
+	Godot::print("Starting passthrough layer...");
+	result = xrPassthroughLayerResumeFB(passthrough_layer);
+	if (!openxr_api->xr_result(result, "Failed to start the passthrough layer")) {
+		stop_passthrough();
+		return false;
+	}
+
+	// Enable the viewport transparent background
+	Viewport *viewport = get_main_viewport();
```

Comment:
In due time we need to improve on this to detect which viewport is actually used. Once desktop headsets start supporting passthrough as well, which they undoubtedly will, it is not guaranteed the main viewport is the one that is used.